File: src/Algorithm/LeetCode/LeetCode_139_Word_Break/Solution_DP.java
package Algorithm.LeetCode.LeetCode_139_Word_Break;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

//https://programmercarl.com/0139.单词拆分.html
//https://leetcode.cn/problems/word-break/solutions/302779/shou-hui-tu-jie-san-chong-fang-fa-dfs-bfs-dong-tai
public class Solution_DP {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> set = new HashSet<>(wordDict);
        boolean[] dp = new boolean[s.length() + 1];

        dp[0] = true;

        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && set.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[s.length()];
    }
}

File: src/Algorithm/LeetCode/LeetCode_139_Word_Break/Solution_DP.py
from typing import List

# https://www.programmercarl.com/0139.单词拆分.html
class Solution:
    def wordBreak(self, s: str, wordDist: List[str]) -> bool:
        wordSet = set(wordDist)
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(1, n + 1):
            for j in range(i):
                if dp[j] and s[j:i] in wordSet:
                    dp[i] = True
                    break

        return dp[n]
